File: pocketpilot-backend/models/monthly_archive.py
from datetime import datetime

from pydantic import BaseModel, Field


class MonthlyArchiveBase(BaseModel):
    """Snapshot of one completed budgeting cycle, created at reset time.

    This is the thing that lets a student glance at "last month" without
    re-running any calculation — the numbers are frozen exactly as they
    were the moment the cycle ended.
    """

    cycle_start: datetime
    cycle_end: datetime
    monthly_budget: float = Field(..., ge=0)
    total_spent: float = Field(..., ge=0)
    total_saved: float
    total_autopays: float = Field(0, ge=0)


class MonthlyArchiveInDB(MonthlyArchiveBase):
    id: str = Field(..., alias="_id")
    user_id: str
    created_at: datetime

    model_config = {"populate_by_name": True}


class MonthlyArchiveResponse(MonthlyArchiveBase):
    id: str
    created_at: datetime

File: pocketpilot_app/lib/models/monthly_archive.dart
class MonthlyArchive {
  final String id;
  final DateTime cycleStart;
  final DateTime cycleEnd;
  final double monthlyBudget;
  final double totalSpent;
  final double totalSaved;
  final double totalAutopays;

  MonthlyArchive({
    required this.id,
    required this.cycleStart,
    required this.cycleEnd,
    required this.monthlyBudget,
    required this.totalSpent,
    required this.totalSaved,
    required this.totalAutopays,
  });

  factory MonthlyArchive.fromJson(Map<String, dynamic> json) => MonthlyArchive(
        id: json['id'],
        cycleStart: DateTime.parse(json['cycleStart']),
        cycleEnd: DateTime.parse(json['cycleEnd']),
        monthlyBudget: (json['monthlyBudget'] as num).toDouble(),
        totalSpent: (json['totalSpent'] as num).toDouble(),
        totalSaved: (json['totalSaved'] as num).toDouble(),
        totalAutopays: (json['totalAutopays'] as num?)?.toDouble() ?? 0,
      );
}
